// @openai/guardrails/dist/types.js
"use strict";
/**
 * Type definitions, interfaces, and result types for Guardrails.
 *
 * This module provides core types for implementing Guardrails, including:
 * - The `GuardrailResult` interface, representing the outcome of a guardrail check.
 * - The `CheckFn` interface, a callable interface for all guardrail functions.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractTokenUsage = extractTokenUsage;
exports.tokenUsageToDict = tokenUsageToDict;
exports.aggregateTokenUsageFromInfos = aggregateTokenUsageFromInfos;
exports.totalGuardrailTokenUsage = totalGuardrailTokenUsage;
const EMPTY_TOKEN_USAGE_SUMMARY = Object.freeze({
    prompt_tokens: null,
    completion_tokens: null,
    total_tokens: null,
});
function isRecord(value) {
    return typeof value === 'object' && value !== null;
}
function isIterable(value) {
    return typeof value === 'object' && value !== null && typeof value[Symbol.iterator] === 'function';
}
function readNumber(value) {
    return typeof value === 'number' && Number.isFinite(value) ? value : null;
}
function pickNumber(record, keys) {
    if (!record) {
        return null;
    }
    for (const key of keys) {
        const candidate = record[key];
        const numeric = readNumber(candidate);
        if (numeric !== null) {
            return numeric;
        }
    }
    return null;
}
/**
 * Extract token usage data from an OpenAI API response object.
 */
function extractTokenUsage(response) {
    const usage = response?.usage;
    if (!usage) {
        return Object.freeze({
            prompt_tokens: null,
            completion_tokens: null,
            total_tokens: null,
            unavailable_reason: 'Token usage not available for this model provider',
        });
    }
    const promptTokens = pickNumber(usage, ['prompt_tokens', 'input_tokens']);
    const completionTokens = pickNumber(usage, ['completion_tokens', 'output_tokens']);
    const totalTokens = pickNumber(usage, ['total_tokens']);
    if (promptTokens === null && completionTokens === null && totalTokens === null) {
        return Object.freeze({
            prompt_tokens: null,
            completion_tokens: null,
            total_tokens: null,
            unavailable_reason: 'Token usage data not populated in response',
        });
    }
    return Object.freeze({
        prompt_tokens: promptTokens,
        completion_tokens: completionTokens,
        total_tokens: totalTokens,
    });
}
/**
 * Convert a TokenUsage object into a plain dictionary suitable for serialization.
 */
function tokenUsageToDict(tokenUsage) {
    const result = {
        prompt_tokens: tokenUsage.prompt_tokens,
        completion_tokens: tokenUsage.completion_tokens,
        total_tokens: tokenUsage.total_tokens,
    };
    if (tokenUsage.unavailable_reason !== undefined) {
        result.unavailable_reason = tokenUsage.unavailable_reason;
    }
    return Object.freeze(result);
}
/**
 * Aggregate token usage values from a collection of guardrail info dictionaries.
 */
function aggregateTokenUsageFromInfos(infoDicts) {
    let totalPrompt = 0;
    let totalCompletion = 0;
    let totalTokens = 0;
    let hasData = false;
    for (const info of infoDicts) {
        if (!info) {
            continue;
        }
        const usage = info.token_usage;
        if (!isRecord(usage)) {
            continue;
        }
        const prompt = readNumber(usage.prompt_tokens);
        const completion = readNumber(usage.completion_tokens);
        const total = readNumber(usage.total_tokens);
        if (prompt === null && completion === null && total === null) {
            continue;
        }
        hasData = true;
        if (prompt !== null) {
            totalPrompt += prompt;
        }
        if (completion !== null) {
            totalCompletion += completion;
        }
        if (total !== null) {
            totalTokens += total;
        }
    }
    if (!hasData) {
        return EMPTY_TOKEN_USAGE_SUMMARY;
    }
    return Object.freeze({
        prompt_tokens: totalPrompt,
        completion_tokens: totalCompletion,
        total_tokens: totalTokens,
    });
}
const AGENT_RESULT_ATTRS = [
    'input_guardrail_results',
    'output_guardrail_results',
    'tool_input_guardrail_results',
    'tool_output_guardrail_results',
    'inputGuardrailResults',
    'outputGuardrailResults',
    'toolInputGuardrailResults',
    'toolOutputGuardrailResults',
];
function extractAgentsSdkInfos(stageResults) {
    if (!stageResults) {
        return [];
    }
    const entries = Array.isArray(stageResults)
        ? stageResults
        : isIterable(stageResults)
            ? Array.from(stageResults)
            : [];
    const infos = [];
    for (const entry of entries) {
        if (!isRecord(entry)) {
            continue;
        }
        const direct = entry.output_info ?? entry.outputInfo;
        if (isRecord(direct)) {
            infos.push(direct);
            continue;
        }
        const output = entry.output;
        if (isRecord(output)) {
            const nested = output.output_info ?? output.outputInfo;
            if (isRecord(nested)) {
                infos.push(nested);
            }
        }
    }
    return infos;
}
/**
 * Unified helper to compute total guardrail token usage from any result shape.
 */
function totalGuardrailTokenUsage(result) {
    if (!isRecord(result)) {
        return EMPTY_TOKEN_USAGE_SUMMARY;
    }
    const guardrailResults = result.guardrail_results ?? result.guardrailResults;
    if (isRecord(guardrailResults)) {
        const totals = guardrailResults.totalTokenUsage;
        if (totals) {
            return totals;
        }
    }
    const directTotals = result.totalTokenUsage;
    if (directTotals) {
        return directTotals;
    }
    const infos = [];
    for (const attr of AGENT_RESULT_ATTRS) {
        const stageResults = result[attr];
        if (stageResults) {
            infos.push(...extractAgentsSdkInfos(stageResults));
        }
    }
    if (infos.length === 0) {
        return EMPTY_TOKEN_USAGE_SUMMARY;
    }
    return aggregateTokenUsageFromInfos(infos);
}
